src/usecases/retrieveAverageParticipantsInVisit.js

```diff
@@ -0,0 +1,54 @@
+import groupBy from "../helpers/groupBy";
+import average from "../helpers/average";
+import { JOIN_VISIT, LEAVE_VISIT } from "../../src/helpers/eventActions";
+
+const retrieveAverageParticipantsInVisit = ({ getDb }) => async (trustId) => {
+  if (!trustId) return { error: "Error: A trustId must be provided." };
+
+  const db = await getDb();
+  const events = await db.any(
+    `SELECT events.id, events.action, events.visit_id, events.session_id
+    FROM events
+    LEFT JOIN scheduled_calls_table ON events.visit_id = scheduled_calls_table.id
+    LEFT JOIN wards ON scheduled_calls_table.ward_id = wards.id
+    LEFT JOIN trusts ON wards.trust_id = trusts.id
+    WHERE trusts.id = $1`,
+    trustId
+  );
+
+  const eventsGroupedByVisit = groupBy(events, "visit_id");
+  const participantsInEachVisit = getParticipantsInEachVisit(
+    eventsGroupedByVisit
+  );
+  const averageParticipantsInVisit = average(participantsInEachVisit);
+
+  return { averageParticipantsInVisit, error: null };
+};
+
+const getParticipantsInEachVisit = (eventsGroupedByVisit) => {
+  let participantsInEachVisit = [];
+
+  for (const [, visitEvents] of Object.entries(eventsGroupedByVisit)) {
+    const eventsGroupedBySesssion = groupBy(visitEvents, "session_id");
+    let participantsInVisit = 0;
+
+    for (const [, sessionEvents] of Object.entries(eventsGroupedBySesssion)) {
+      const joinEvent = sessionEvents.find(
+        (event) => event.action === JOIN_VISIT
+      );
+      const leaveEvent = sessionEvents.find(
+        (event) => event.action === LEAVE_VISIT
+      );
+
+      if (joinEvent && leaveEvent) {
```

Comment:
Do we need to check for both join and leave visit events? Have we considered simply counting the unique session IDs grouped by visit ID and then getting the average count from there ?

Reply:
Nope, I didn't consider just using the session ID! I did consider grouping in the SQL statement like I said in the Notes but as a first pass I decided to write the simplest SQL statement first (as it's not something I'm very knowledgeable about so didn't want to get bogged down on perfecting a statement) and use JS to do the rest.

If doing things in Postgres is the way we want to go, could we 🍐 on it?

Reply:
I think it turned out easier than I thought! Just pushed up https://github.com/madetech/nhs-virtual-visit/pull/455/commits/e03f0f20c13b09170b8b6490dbb8b06947915f1a, let me know if this is what you were thinking.